tools/resource-gen/main.go

```diff
@@ -418,34 +407,127 @@ func main() {
 		resources = append(resources, resourceInfo)
 	}
 
-	var generatorTemplate *template.Template
+	var generatorFn GeneratorFn
 
 	switch gen {
 	case "type":
-		generatorTemplate = ResourceTemplate
+		generatorFn = TemplateGeneratorFn(ResourceTemplate)
 	case "crd":
-		generatorTemplate = CustomResourceTemplate
+		generatorFn = TemplateGeneratorFn(CustomResourceTemplate)
+	case "openapi":
+		generatorFn = openApiGenerator
 	default:
 		log.Fatalf("%s is not a valid generator option\n", gen)
 	}
 
-	outBuf := bytes.Buffer{}
-	if err := generatorTemplate.Execute(&outBuf, struct {
-		Package   string
-		Resources []ResourceInfo
-	}{
-		Package:   pkg,
-		Resources: resources,
-	}); err != nil {
-		log.Fatalf("template error: %s", err)
-	}
-
-	out, err := format.Source(outBuf.Bytes())
+	err := generatorFn(pkg, resources)
 	if err != nil {
 		log.Fatalf("%s\n", err)
 	}
+}
 
-	if _, err := os.Stdout.Write(out); err != nil {
-		log.Fatalf("%s\n", err)
+func openApiGenerator(pkg string, resources []ResourceInfo) error {
+	protoTypeToType := map[string]reflect.Type{
+		"Mesh":        reflect.TypeOf(v1alpha1.Mesh{}),
+		"MeshGateway": reflect.TypeOf(v1alpha1.MeshGateway{}),
+	}
+
+	for _, r := range resources {
+		tpe, exists := protoTypeToType[r.ResourceType]
+		println("generating", r.ResourceType)
+		if !exists {
+			continue
+		}
+		reflector := jsonschema.Reflector{
+			ExpandedStruct:            true,
+			DoNotReference:            true,
+			AllowAdditionalProperties: true,
+		}
+		err := reflector.AddGoComments("github.com/kumahq/kuma/", "api/")
+		if err != nil {
+			return err
+		}
+		schemaMap := orderedmap.New[string, *jsonschema.Schema]()
+		schemaMap.Set("type", &jsonschema.Schema{Type: "string"})
+		schemaMap.Set("name", &jsonschema.Schema{Type: "string"})
+		if !r.Global {
+			schemaMap.Set("mesh", &jsonschema.Schema{Type: "string"})
+		}
+		schemaMap.Set("labels", &jsonschema.Schema{Type: "object", AdditionalProperties: &jsonschema.Schema{Type: "string"}})
+		properties := reflector.ReflectFromType(tpe).Properties
+		for pair := properties.Oldest(); pair != nil; pair = pair.Next() {
+			schemaMap.Set(pair.Key, pair.Value)
+		}
+
+		schema := jsonschema.Schema{
+			Type:       "object",
+			Required:   []string{"type", "name", "spec"},
```

Comment:
```suggestion
			Required:   []string{"type", "name"},
```

Reply:
Also you should add a: `if !r.Global { schema.Required = append(schema.Required, "mesh")}`